kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java

```diff
@@ -0,0 +1,417 @@
+     */
+    @Override
+    public CompletionStage<VirtualClusterBinding> resolve(String bindingAddress, int port, String sniHostname, boolean tls) {
+        var endpoint = new Endpoint(bindingAddress, port, tls);
+        var lcr = this.listeningChannels.get(endpoint);
+        if (lcr == null || lcr.unbindingStage().get() != null) {
+            return CompletableFuture.failedStage(buildEndpointResolutionException("Failed to find channel matching ", bindingAddress, port, sniHostname, tls));
+        }
+        return lcr.bindingStage().thenApply(channel -> {
+            var bindings = channel.attr(CHANNEL_BINDINGS);
+            if (bindings == null || bindings.get() == null) {
+                throw buildEndpointResolutionException("No channel bindings found for ", bindingAddress, port, sniHostname, tls);
+            }
+            // We first look for a binding matching by SNI name, then fallback to a null match.
+            var binding = bindings.get().getOrDefault(RoutingKey.createBindingKey(sniHostname), bindings.get().get(RoutingKey.NULL_ROUTING_KEY));
+            if (binding == null) {
+                throw buildEndpointResolutionException("No channel bindings found for ", bindingAddress, port, sniHostname, tls);
+            }
+            return binding;
+        });
+    }
+
+    private EndpointResolutionException buildEndpointResolutionException(String prefix, String bindingAddress, int port, String sniHostname, boolean tls) {
+        return new EndpointResolutionException(
+                ("%s binding address: %s, port %d, sniHostname: %s, tls %s").formatted(prefix,
+                        bindingAddress == null ? "any" : bindingAddress,
+                        port,
+                        sniHostname == null ? "<none>" : sniHostname,
+                        tls));
+    }
+
+    /**
+     * Indicates if the registry is in closed state.
+     * @return true if the registry has been closed.
+     */
+    public boolean isRegistryClosed() {
+        return registryClosed.get();
+    }
+
+    /**
+     * Signals that the registry should be shut down.  The operation returns a {@link java.util.concurrent.CompletionStage}
+     * that will complete once shutdown is complete (endpoints unbound).
+     *
+     * @return CompletionStage that completes once shut-down is complete.
+     */
+    public CompletionStage<Void> shutdown() {
+        return allOfStage(registeredVirtualClusters.keySet().stream().map(this::deregisterVirtualCluster).toList())
+                .whenComplete((u, t) -> {
+                    LOGGER.debug("EndpointRegistry shutdown complete.");
+                    registryClosed.set(true);
+                });
+    }
+
+    @Override
+    public void close() throws Exception {
+        shutdown().toCompletableFuture().get();
+    }
+
+    private static <T> CompletableFuture<Void> allOfStage(Collection<CompletionStage<T>> futures) {
```

Comment:
nit: Maybe change this to accept a stream of stages rather than collecting and re-streaming.